sklearn/utils/metaestimators.py

```diff
@@ -111,7 +111,12 @@ def __get__(self, obj, owner=None):
                 )
 
         # lambda, but not partial, allows help() to work with update_wrapper
-        out = lambda *args, **kwargs: self.fn(obj, *args, **kwargs)  # noqa
+        if obj:
+            out = lambda *args, **kwargs: self.fn(obj, *args, **kwargs)  # noqa
+        else:
+            # This makes it possible to use the decorated method as an unbound method,
+            # for instance when monkeypatching.
+            out = lambda *args, **kwargs: self.fn(*args, **kwargs)  # noqa
```

Comment:
@ogrisel 
For https://github.com/scikit-learn/scikit-learn/pull/20623#discussion_r678132317, should we also perform the attribute check for the unbound case?


```suggestion
            # This makes it possible to use the decorated method as an unbound method,
            # for instance when monkeypatching.
            def out(*args, **kwargs):
                obj = args[0]
                if not self.check(obj):
                raise AttributeError(
                    f"This {repr(owner.__name__)}"
                    " has no attribute"
                    f" {repr(self.attribute_name)}"
                )
                return self.fn(*args, **kwargs)
```

Reply:
emm, i rethink on this, seems no need, when we call the function, it will delegate to target function first and then raise related error. withdraw my old comment.

Reply:
Yes, under the assumption that there will still be an error raised, it should make negligible difference whether or not we run the available_if check. If we expect some level of consistency between using the bound vs unbound method, however, we could consider running the check and raising an AttributeError on failure...